### ISSUE TITLE
Issue with large DB

### ISSUE THREAD

**Issue by Robert99943** [mserajnik/hydrusrv#3]
I have a Hydrus server with 209,174 files and 20,594,869 total mappings for tags. I just uploaded all my local tags I had cause there isn't really a good way of mass uploading tags otherwise.
DB's files total up to about 1.1GB.

When trying to run hyrdusrv with this db selected in .env file, I cannot connect to the service. Before, when it didn't have quite as many files but same amount of tags, it would just seemingly load indefinitely but now it just doesn't connect.

**Comment by mserajnik** [comment]
While it‘s a bit odd that you can‘t access the API at all (which should be available regardless of what‘s happening with the data in the background), I‘m pretty sure it‘s the size of the database that‘s causing the issues.

If I had to guess, it would either be the amount of mappings itself (I‘m pulling the whole thing into memory) or my `WHERE IN` checks that limit the mappings I select to match the files/tags it has selected before (which in hindsight isn‘t very smart if we are dealing with 10k+ of each).

I will probably need to rewrite this whole data handling (introduce chunking, use smarter queries etc.) for it to work with large databases like yours. But I had planned to do this at some point anyway.

Is there a chance I could get a copy of your database(s) (just the DB files, not the actual media files)? It would make working on this a lot easier since I don‘t have any database that size myself and setting one up via downloading actual files or even writing a tool to generate a „mocked“ one would take quite some time.

**Comment by mserajnik** [comment]
FYI, I've just noticed a pretty big oversight: https://github.com/mserajnik/hydrusrv/blob/master/server/helpers/data.js#L342-L343

This gets slow really fast with an increasing amount of files. I've refactored it locally so it only runs a single query per namespace (and do the rest in code) and was able to reduce the time an update took on my [test instance](http://hydrusrvue.mser.at) by half (from 4 seconds to 2 seconds). This should be even (far) more severe in larger sets like yours, it's likely the biggest offender by a large amount.

I'm going to try some more optimizations and push an update today or later this week.

**Comment by Robert99943** [comment]
Thanks for the help. After leaving it running overnight I do seem to be able to access the API, but it still seems to load indefinitely. I will try letting this load for a while and see if it does eventually load.

Here are my DB files: https://koto.reisen/db.7z

**Comment by mserajnik** [comment]
Thank you for the database files, this helped __a lot__. :)

After spending the evening doing some more optimization and a lot of testing I was able to get the update time down to about 430 seconds using your database (with `HYDRUS_SUPPORTED_MIME_TYPES` set to everything hydrusrv supports). On my local machine and outside of Docker mind you, so it will likely be a bit slower when running on a server and/or dockerized.

Still, this should hopefully be fine for everyday use, at least if you don't plan on running data updates every 30 minutes or so (personally, I only run updates once a day). I still need to test this to make sure I didn't introduce any bugs before making a new release, but if you want to check it out, it's on the `develop` branch: https://github.com/mserajnik/hydrusrv/commit/43cb341915a8425d0ae0c5e1b9fc8d1d79d8fdb4 (I'll probably amend and force-push the commit quite a bit, so the link might not be the correct one in future revisions)

One thing I did notice is that only about 138k files, 71k tags and 4.3m mappings are actually used by hydrusrv – do you perchance have a lot of unsupported file types that are not in [this list](https://github.com/mserajnik/hydrusrv/blob/master/server/config/hydrus.js#L1-L15)? Otherwise, your database might have some issues and/or we just found another bug in hydrusrv. :D

Oh, and the API not being available before the first update is finished is normal – that's why it didn't work for you (because the update probably ran for several hours). I forgot about before.

I sadly cannot replicate the indefinite loading you describe; I assume you're using [hydrusrvue](https://github.com/mserajnik/hydrusrvue) to access the API? Are there any errors in the console or can you see if the requests actually finish in the network tab?

**Comment by Robert99943** [comment]
All the files on the server are listed as supported, in fact I believe they are all either jpg, png, gif, mp4, or webm. I have tried both hydrusrvue to access and just the IP. I tried leaving it to load on the IP since my last message so about 8 hours and it never did. The request to the API in the network tab doesn't finish and the console doesn't give any errors.

**Comment by Robert99943** [comment]
I think I might have found what was causing the indefinite loading, I had a process of `node ./bin www` running and I couldn't kill it. I rebooted and switched to the develop branch and I can access the API just fine. I do notice that hydrusrvue doesn't show all the 209k files as well, although It says 159831 files.

**Comment by Robert99943** [comment]
Also, possibly related to the not showing all files. A good bit of images show the wrong tags on hydrusrvue.

![fzdql](https://user-images.githubusercontent.com/15913135/46638900-23d81580-cb31-11e8-887d-b5402750c931.png)

It seems to be showing the tags of another image, but not all of them seem to have all the exact tags from another image.

If you wanted to see.
https://booru.koto.reisen/#/files/62767 
login `guest:guest123` (some NSFW files in the repo)

**Comment by mserajnik** [comment]
Thanks for the feedback. I'm glad at least the performance improvements seem to help. :)

It's weird that you would have a zombie `node` process, hydrusrv should shut down correctly on `SIGTERM`/`SIGKILL` (so, also by hitting `Ctrl+C` or closing the terminal session).

Regarding the missing files/wrong tags: these issues did pop up for me earlier in development, but I had frankly hoped they would've been solved by now. To explain in a bit more detail:

__Beware, wall of text!__

---

hydrus network uses repositories to store tags and files. Each repository has its own service ID (usually `2` for the first tag repository on a server and `3` for the first file repository, although I've noticed they are switched in your database) which is then used to suffix database tables, like `repository_hash_id_map_3` or `current_mappings_2`.

Now, in general, everything tag-related is stored in tables suffixed with `2` (assuming `2` is the service ID of our tag repository) and everything file-related in tables suffixed with `3`. The exception seem to be the tables called `repository_hash_id_map_x`, which basically map files from the "master tables" (which hold the meta information of every file on the server instance) to the services.

What is weird is that both `repository_hash_id_map_2` and `repository_hash_id_map_3` can contain different and/or (some of) the same mappings, depending on how files and their tags were uploaded.

Since I obviously don't want to have any duplicates, I first get all files from the `repository_hash_id_map_x` table with the service ID of the tag repository (to get all files that have tags) and then get all the files from the `repository_hash_id_map_x` table with the service ID of the file repository, excluding any files that I already have to end up with no duplicates.

According to the developer of hydrus himself, the mappings should always point to the same files on both tables. E.g., if a mapping exists on the tag repository table with the service-internal ID `42` that points to the master ID `100`, another mapping on the file repository table with the same service-internal ID (`42`) should also point to the same master ID (`100`). So it _shouldn't_ make any difference which mapping you use.

But from my testing (and also from what you are seeing right now), this doesn't seem to always be the case. The same service-internal IDs on both tables _can_ point to different master IDs, for whatever reason. And apparently, one could only guess which is the correct one to use for every file to end up with no missing files and correctly assigned tags in all cases. I've also tried going by the timestamp field and always using the older or newest one, but that also doesn't seem to help.

Weirdly, I myself stopped running into these issues after adjusting the following things:

- Make hydrusrv not lock the database of hydrus server when reading; I had initially done this to prevent data modification during updates, but since hydrus server certainly doesn't expect a locked database, I figured it would be better to check for such modifications on my end and handle them accordingly instead of denying hydrus server write access
- Always shut down hydrus server gracefully; this was mainly an issue with my Docker setup where I would just kill the container instead of sending a `SIGTERM` to hydrus server first
- Use non-[WAL](https://www.sqlite.org/wal.html) mode when running hydrus server; this just seemed to make sense since I was restarting the server quite often during development and wasn't sure if that didn't cause any broken transactions or anything

Now, I'm not sure if anything of this had an effect, but like I said, those issues were gone for me after these changes.

I'm frankly a bit out of ideas and given the fact that even the hydrus developer himself told me the `repository_hash_id_map_x` mappings should always point to the same master IDs makes me believe this might be a bug in hydrus server itself – one that might not manifest itself when accessing it with the hydrus client because hydrus handles data a lot differently than I do (due to not being constrained by Web limitations, not needing to paginate results, having support for multiple repositories etc.).

Sorry for the long text but I figured I might as well post this explanation here since it's relevant to the issues you are experiencing.

---

__TLDR:__ hydrus server shows some (to me) weird database behavior that I can't really understand and from the feedback I got from its developer I'm concluding that it might even be a bug in hydrus server itself.

I would suggest you try out my [Docker setup](https://github.com/mserajnik/hydrusrv-docker), if you can – just to see if that makes any difference.

I'm also open for suggestions and/or help. I've talked with several colleagues/friends about those issues and no one was able to come up with a solution or even an explanation as to what exactly is happening. :S

**Comment by mserajnik** [comment]
I'll try to take a look at this this weekend, starting from scratch in regards to how I read from the hydrus server databases.

Maybe there's something fundamentally wrong with how I select and try to to merge data from the two repositories and I need a different approach.

**Comment by mserajnik** [comment]
So, I've had another good look at this and frankly, I don't think I'm doing anything wrong. I've checked the example you've provided (with ID `62676`) manually and the tags hydrusrv assigns are the correct ones according to the relations present in the database.

Now, since they are clearly _not_ the correct ones and you're also missing a huge amount of files (despite them having supported MIME types), my best bet would be the following:

If you are using WAL mode (which is the default hydrus server operates with) and have terminated hydrus server at any point while there were still pending database changes, those changes might've been lost, which might in turn have led to those missing files and wrongly assigned tags (because some IDs had already been updated while others hadn't). Assuming hydrus server does quite a lot of data rearranging to optimize performance and keep everything clean (at least from what I've seen), this might've even happened at a later point after you had already finished inserting the data.

So, if you want to, I'd recommend doing the following:

- Create a new hydrus server instance (I'd start fresh to eliminate the possiblity of anything going wrong due to old data)
- Make sure you start hydrus server using non-WAL mode (via `./server --no_wal`)
- Make sure to stop hydrus server "cleanly" using `Ctrl+C` or `./server stop`
- Upload all your files and tags and give it a try

Alternatively, you can use [hydrusrv-docker](https://github.com/mserajnik/hydrusrv-docker) which also uses non-WAL mode and tries to shut down hydrus server in the recommended way when you stop the container.

As mentioned before, I haven't had any issues after with wrong tags/missing files after doing these "optimizations" on my end.

Aside from this, I'm pretty much out of ideas though, so if this doesn't help, I can only recommend using a different software (like [szurubooru](https://github.com/rr-/szurubooru) for example) for the time being. I'll still make a new release soon that includes the performance improvements I've added.

**Comment by FredericaBernkastel** [comment]
Kinda related.
Installed latest dev branch (8cdcf10) on ubuntu 18.04.
With few files and tags it worked well. Then i pushed 1M tag mappings, gracefully stopped hydrus, and ran hydrusrv.
It stuck for 20 minutes (100% cpu load), api was unaccessible.

Ok, gracefully stopped hydrusrv, pushed few images to check that new mappings actually work.
Hydrusrv stuck again. Is it normal? It doesnt cache anything?

After another 20 minutes, i ran hydrusrvue, and noticed that all tags are **wrong** (randomly shuffled between images).

**Comment by FredericaBernkastel** [comment]
Also, it crashes in dev mode.
```bash
$ yarn start-dev
yarn run v1.12.1
$ nodemon ./bin/www
(node:8626) [DEP0106] DeprecationWarning: crypto.createDecipher is deprecated.
[nodemon] 1.18.6
[nodemon] to restart at any time, enter `rs`
[nodemon] watching: *.*
[nodemon] starting `node ./bin/www`
[nodemon] Internal watch failed: ENOSPC: System limit for number of file watchers reached, watch '/home/bernkastel/.hydrusrv/node_modules/uri-js/src/util.ts'
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by mserajnik** [comment]
@FredericaBernkastel 

Thanks for the report, sorry that this causes you trouble. :/

1m mappings really shouldn't take that long to complete. How many files/tags do you have? @Robert99943 had over 20m/200k files and after doing some improvements I got the update time down to about 430 seconds (I don't even know how long it took before, but it was certainly over 45 minutes, that's the longest I've let it run). Still not ideal, but if you only run updates in hydrusrv once a day or so it shouldn't be that big of an issue.

But this definitely sounds like either a bug or a general performance issue somewhere in the code. If you want to test this yourself, I usually put `console.time()`/`console.timeEnd()` calls (see [here](https://developer.mozilla.org/en-US/docs/Web/API/Console/time)) into [this area](https://github.com/mserajnik/hydrusrv/blob/master/server/helpers/data.js#L18-L70) in meaningful places  to check how long reading and writing the data takes. It should become obvious pretty fast what's taking so long here. Otherwise, I can also offer you to take a look at it using your DB files, but I'm not sure how soon I could get to that.

No, unfortunately, it doesn't cache anything at the moment. All the data is put into temporary tables that only live until the next update (or restart). I've tried comparing data and only updating when there were changes in hydrus before, but this turned out to be rather difficult. Basically, I couldn't rely on IDs staying the same in hydrus since it looked like the automatic cleanup it does causes files/tags getting assigned new ones from time to time.

I'm considering tackling this again and have some ideas how to approach it in a sane way, but I didn't want to start this until the more pressing issue (wrongly assigned tags) is fixed.

---

So, as I've stated in here before, I'm a bit out of ideas as to how to address this tags issue. I'm pretty sure I know _what_ is happening though (I'll be repeating a lot of what I said before here, but I hope this is a more condensed explanation of what's going wrong):

The `service_hash_id` column relates a file to its tags and while it should always use this ID from the tag repository related mappings table (since all the service hash IDs from files having tags should be stored in there), it looks like that it sometimes hydrus uses the file repository related mappings table to store IDs for files with tags instead. And afaik, that shouldn't happen, only files without tags should have their service hash IDs stored in there.

And since the same service hash ID in the two tables (let's say `42`) can – even though according to the hydrus developer they shouldn't – apparently point to different master hash IDs (which are used to get to the actual file and its metadata like width, height, size etc.), these mixups can occur and cause tags to appear on the wrong files, even though it's the correct service hash ID (just from the wrong table).

Now, this doesn't seem to be a problem at all for hydrus itself, so I'll probably have to adjust how I'm reading the files from the DB. This all happens [here](https://github.com/mserajnik/hydrusrv/blob/master/server/helpers/data.js#L296-L400), by the way.

I'm not sure when I'll find the time to do that, so I'm glad for any help.

Other than that, I haven't encountered this issue on my personal DB since before I did the adjustments I've mentioned in a previous post (non-WAL mode etc.), but that might've been coincidence.

---

Thanks for letting me know that `yarn start-dev` doesn't work for you, I can see why it would fail if it tried to watch the whole `node_modules` directory (looks like it from the error message). It shouldn't do that per default, but this has probably changed in a recent update. Or at the least, it never failed for me before, might also be a limit set on OS level on Ubuntu/Debian that didn't come up on my macOS dev machine.

In any case, I'll try to get a hotfix up for this dev mode issue in the coming days.

**Comment by FredericaBernkastel** [comment]
Ye, it's a known issue.
https://github.com/remy/nodemon/issues/339

I'll dig into your code, once i get a free time. Let's see what we can do.
At this point, it's not ready for production, even for moderate-sized collections.

**Comment by mserajnik** [comment]
![image](https://user-images.githubusercontent.com/28833155/48097006-fefebd00-e218-11e8-87fd-701524b39231.png)

I found this screenshot I sent to the hydrus developer a few months ago which illustrates the issue with the tags being assigned to the wrong master hash ID (= the wrong file in the end) pretty well I think. Basically, according to the hydrus developer, the master hash ID should be the same in both `repository_hash_id_map_x` tables for a given service hash ID.

I've asked him about this not being the case here and his reponse was:

> Yeah, the master_hash_id should be the same for the same file, so if you applied tags to the 52KB file with master_hash_id 71, then the tag repo should be assigning it to master_hash_id 71. It only seemed correct to me because in your previous screenshots it was 1/71 and 71/71, which seemed all correct. I cannot explain how it would line up the wrong master_hash_id.

So there definitely seems to be some kind of bug or oversight in hydrus itself (that seemingly doesn't affect it though) – I just figured I had fixed it with the changes I've made before. Apparently not. So I'll need to see if I can work around that (probably by figuring out how hydrus server reads its data when creating updates for the client, I haven't quite got the hang of that yet).

Thanks for the offer to help, I appreciate it. I'll also try to get some free time to look at it.

Well, I (wrongly) _assumed_ it was ready since my testing with around 10k files/30k tags/150k mappings didn't bring up issues anymore. Given that I personally don't horde huge amount of files, I didn't feel the need to test with millions of files/mappings. That's also why I added the [caveats section](https://github.com/mserajnik/hydrusrv#caveats).

I just thought it was a neat idea that hasn't been done before (basically adding a layer on top of hydrus server to serve its files outside of hydrus client updates) and that somebody might find a use case for, so again, I'm really sorry it doesn't work as expected. :(

Honestly, had I known how much trouble this would cause, I'd have gone with my initial plan, which would basically be a standalone booru server without the need for hydrus server or even any kind of administration interface. More like cloud storage like Dropbox where you drop files/tags files (<- should work with hydrus client exports) into a media directory and it does everything in the background (inserting into the DB, generating thumbnails, converting videos etc.) and serves the files/tags via API and has some basic user management, just like hydrusrv does.

This could then be paired with a little helper tool running alongside a hydrus client that rsync's hydrus exports to the server (instead of running a hydrus server and uploading to that one) to get the same functionality out of it as this one. Well, I might still have a go at that one if I can't find a satisfying solution for these issues here. :D

**Comment by mserajnik** [comment]
I just realized, we have timestamps on the `current_files_x` (only the file repository table is used here) and both the `repository_hash_id_map_x` tables.

Instead of determining the newest of two timestamps for a given service hash ID that exists in both `repository_hash_id_map_x` tables like I tried before (didn’t work reliably; apparently, newest doesn‘t always mean it‘s also the one used), I think it might work the following way instead:

1. Get all entries from `current_files_x`
2. For every service hash ID, check if it exists in either `repository_hash_id_map_x` table
  + If it only exists in one, just use the master hash ID from that entry
  + If it exists in both, compare the timestamp from the `current_files_x` entry to the one on the entries in both `repository_hash_id_map_x` tables -> it should be clear which one is the correct one, because the correct one was probably either inserted right before or after (need to check the hydrus code to find out) the entry in `current_files_x`

If this works, it should completely get rid of the tags issue. It‘s a rather simple change as well, I might find the time to try this later today.

**Comment by mserajnik** [comment]
**Update:** Scratch my last post. I've just had a realization. Don't want to go into details yet in case I'm mistaken, but I think I might've found the issue. Looks like I'm doing something fundamentally wrong when selecting the files. I'll give it a try and update once I know more.

**Comment by mserajnik** [comment]
@FredericaBernkastel 

Alright, this looks pretty promising. Could you try using the following branch and see if it still assigns the wrong tags? https://github.com/mserajnik/hydrusrv/tree/hotfix/3.12.2 (if you can work around the performance issues you have – sorry, no fix for that yet).

It seems to work fine in my local tests and [the demo](https://hydrusrvue.mser.at/#/).

Basically, what I've done is that the service hash ID from the files repository is now always used to identify the file and the service hash ID from the tags repository is always used to get any tags, so I store both. Before, it would use either/or for both tasks, depending on if a file has tags or not (which until now seemed to make sense in my head).

Looking at it now, I'm really not sure why I didn't try to do it like this from the beginning, it makes far more sense as well and I'm hopefully optimistic that this might be the fix.

**Careful:** due to hydrusrv now always using the service hash ID from the files repository to identify files, they likely (unless the service hash ID from the tags repository is the same by chance) won't be accessible under their old URLs when using this branch (e.g., `/files/530` will likely return a different file than before).

**Comment by mserajnik** [comment]
Great, let me know if they are still correct once you push those 20k files. :)

5 files and 695k mappings? So you have a huge amount of of different tags on just those few files? Or are these your mappings for all your 20k files?

I'm curious what exactly is taking so long in the update for you. But I can probably immediately improve performance a bit by removing this bit (and some related code): https://github.com/mserajnik/hydrusrv/blob/hotfix/3.12.2/server/helpers/data.js#L400-L404

If I didn't overlook something, this shouldn't be needed anymore since it's already correctly limited to the current files/tags by the updated query. Still, best to confirm the tags issue is indeed gone before I'm going to make more changes here. :)

Other than that, there's probably still quite a bit that can be done to speed up the update process. I'm also thinking about switching from temporary tables to just normal tables, then it would at least have some sort of caching in case a server restart is needed. Could be paired with a flag that disables the initial update when starting the server so the first update runs after the configured time period. This would be quick to implement and wouldn't require any major rewrites.

**Comment by FredericaBernkastel** [comment]
Pushed 20k files, been waiting for 4 hours to finish update, got bored and just killed the process.

Then i decided to do some tests. Made new repos from scratch, pushed 4.5k files / 173k mappings, and noticed inadequate RSS - 533MB!

![clipboard-75](https://user-images.githubusercontent.com/16236235/48179796-0c1bc900-e34b-11e8-8abd-377c78976b26.png)
.

**Comment by mserajnik** [comment]
I've amended the commit on https://github.com/mserajnik/hydrusrv/tree/hotfix/3.12.2 and added the changes where I removed the unnecessary portions mentioned in my last comment.

Hm, the RAM issue seems plausible, although when testing @Robert99943's DB locally on my dev machine (which has 16GB, so plenty available to allocate), it never went far beyond 1GB for the Node processes iirc.

So, I wouldn't expect such a huge difference (40h for you vs around 7 minutes for his DB with 20m mappings and 200k files) between let's say 500MB and 1GB available for Node. It's surely a factor, but I'm guessing there's also some difference in data (maybe a huge amount of namespaces in your DB?) that's causing these discrepancies.

**Comment by mserajnik** [comment]
You can check the error message by putting a `console.error(err)` (or something like that) inside [this `catch` block](https://github.com/FredericaBernkastel/hydrusrv/blob/master/server/helpers/data.js#L45-L56).

The only thing I can think of is that hydrus had some pending files or tags that were finally added between the update task reading the files/tags and the mappings. Because unless the repositories really weren't finished being set up at that point or something completely unlikely like reading from the DB went wrong, inserting the mappings is pretty much the single point of error here.

This would then cause a foreign key constraint error when inserting the mapping (in case a file/tag with the ID from a mapping doesn't exist). This was alleviated by the code I've now removed (which basically checked all the mappings against the already fetched files/tags and remove the ones that don't have pairings) but I hoped that since it correctly assigns the tags for every file now, it wouldn't be needed anymore. Will need to take a closer look at that, probably missed something.

Thanks for the extensive testing in any case, if you could provide me the last working test DB and the ones that doesn't work, this would be great. In addition, if you want, I can also take a look at the DB where the update took forever for you.

Won't have too much time the coming days, but I'll probably find some starting Monday evening again.

**Comment by FredericaBernkastel** [comment]
```
createTempFilesTable...
SqliteError: near "(": syntax error
    at Object.createTempFilesTable (/home/bernkastel/.hydrusrv/server/helpers/data.js:183:12)
    at Object.sync (/home/bernkastel/.hydrusrv/server/helpers/data.js:49:12)
    at updateData (/home/bernkastel/.hydrusrv/app.js:26:10)
    at Object.<anonymous> (/home/bernkastel/.hydrusrv/app.js:34:1)
    at Module._compile (internal/modules/cjs/loader.js:707:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:718:10)
    at Module.load (internal/modules/cjs/loader.js:605:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:544:12)
    at Function.Module._load (internal/modules/cjs/loader.js:536:3)
    at Module.require (internal/modules/cjs/loader.js:643:17) 'error: near "(": syntax error'
```

What if namespace has "( )" chars?

**Comment by mserajnik** [comment]
Should actually be discarded in that case (the whole namespace), as mentioned in the readme:

> hydrusrv discards namespaces containing other characters than alphanumerics and underscores to not falsely assume emote tags like `>:)` have a namespace and to prevent errors (namespaces are used as column names and SQLite does not allow most characters aside from alphanumerics in those; mapping such special characters would have been possible, but did not seem worth the effort).

But that's clearly a bug. Could you `console.log(namespaces)` after [this line](https://github.com/FredericaBernkastel/hydrusrv/blob/master/server/helpers/data.js#L22) to see if it fetched one that's not supported?

**Comment by FredericaBernkastel** [comment]
[namespaces2.txt](https://github.com/mserajnik/hydrusrv/files/2563555/namespaces2.txt)

![clipboard-76](https://user-images.githubusercontent.com/16236235/48226812-d9191a00-e3ca-11e8-9dc1-c7f17d951f52.png)

:joy:

**Comment by mserajnik** [comment]
Thanks, just looking at it, it's likely the second `:` in there that I didn't envision. xD

Regarding my previous post (edited): would you prefer it to keep all the namespaces and just replace unsupported characters instead of dropping them altogether?

I think I could add an environment variable to exclude emotes from the namespaces which I would set defaults for. I could determine these defaults by getting all the emotes that contain `:` from Danbooru (or other popular boorus). Users could then add additional ones to exclude if they want.

**Comment by FredericaBernkastel** [comment]
Right, emotes should be whitelisted (as i saw in other booru engines).
and bad namespaces like this will be discared, but log them to console (to let user decide what to do with)

**Comment by mserajnik** [comment]
I'll fix this and just log the discarded namespaces for now (good suggestion), maybe make it configurable in a future feature release (let the user decide if he wants to discard it or just replace unsupported characters and allow him to exclude tags from being recognized as having namespaces).

By the way, this double `:` error, that's the issue you've had with your 18k files test DB right? Because if that's the case, I'd assume I can keep the last changes for now and don't have to adjust or revert them.

I'll try to get the hotfix release for the wrongly assigned tags and the double `:` issue out tomorrow and do some more performance testing next week. 👍

**Comment by mserajnik** [comment]
Made a regular release instead and updated some dependencies while I was at it: [`3.13.0`](https://github.com/mserajnik/hydrusrv/tree/3.13.0)

No logging of discarded namespaces yet, sorry. I realized this would require a bit more work than I anticipated and I'll probably tackle this when adding the mentioned configuration options instead.

**Comment by FredericaBernkastel** [comment]
Also, please add colors config/preset for common namespaces!
Otherwise each new registered user will be forced to configure them by himself.

**Comment by mserajnik** [comment]
> Also, please add colors config/preset for common namespaces!

I think that's a good idea, but that's something to do in hydrusrvue (I'll open a new issue there and assign it to myself) – I really want hydrusrv to be just the server/API that doesn't care about such things. That's why it only gets saved in local storage in the first place. So I'll add presets that get set not on a per user, but a per browser/device basis to fit with that philosophy.

I'll leave this issue open for now in case @Robert99943 wants to chime in if the latest release also fixed the issues regarding wrongly assigned tags for him.

For additional problems (e.g., performance issues we haven't already talked about), please open new issues – this thread is getting kinda long. 😅

**Comment by Robert99943** [comment]
Sorry for not getting back to this earlier, I tried a non WAL db and it didn't work and had some other issues with docker, so I kind of just gave up. I am currently trying the new version, only 2 million tags committed so far but tested out a few dozen files and I have not seen any incorrectly assigned tags.

So thanks and nice work @mserajnik, and thanks @FredericaBernkastel for reviving this issue.

**Comment by mserajnik** [comment]
@Robert99943 Np, thanks for letting me know. 👍

I'll close this for now, feel free to reopen (or open a different issue) if this comes up for you again.
